azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -736,32 +731,159 @@ def do_retry():
+
+            else:
+                logger.info(
+                    "{}({}): State is {}.  Sending op down.".format(self.name, op.name, self.state)
+                )
+                self.send_op_down(op)
 
         else:
             self.send_op_down(op)
 
     @pipeline_thread.runs_on_pipeline_thread
-    def _set_reconnect_timer(self):
+    def _handle_pipeline_event(self, event):
+        if isinstance(event, pipeline_events_base.DisconnectedEvent):
+            if self.pipeline_root.connected:
+                logger.info(
+                    "{}({}): State is {}.  Triggering reconnect timer".format(
+                        self.name, event.name, self.state
+                    )
+                )
+                self.state = ReconnectState.WAITING_TO_RECONNECT
+                self._start_reconnect_timer()
+            else:
+                logger.info(
+                    "{}({}): State is {}.  Doing nothing".format(self.name, event.name, self.state)
+                )
+
+            self.send_event_up(event)
+
+        else:
+            self.send_event_up(event)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _send_new_connect_op_down(self):
+        self_weakref = weakref.ref(self)
+
+        @pipeline_thread.runs_on_pipeline_thread
+        def on_connect_complete(op, error):
+            this = self_weakref()
+            if this:
+                if error:
+                    if this.state == ReconnectState.NEVER_CONNECTED:
+                        logger.info(
+                            "{}({}): error on first connection.  Not triggering reconnection".format(
+                                this.name, op.name
+                            )
+                        )
+                        this._complete_waiting_connect_ops(error)
+                    elif type(error) in transient_connect_errors:
+                        logger.info(
+                            "{}({}): State is {}.  Connect failed with transient error. Triggering reconnect timer".format(
+                                self.name, op.name, self.state
+                            )
+                        )
+                        self.state = ReconnectState.WAITING_TO_RECONNECT
+                        self._start_reconnect_timer()
+
+                    elif this.state == ReconnectState.WAITING_TO_RECONNECT:
+                        logger.info(
+                            "{}({}): non-tranient error.  Failing all waiting ops.n".format(
+                                this.name, op.name
+                            )
+                        )
+                        self.state = ReconnectState.CONNECTED_OR_DISCONNECTED
+                        self._clear_reconnect_timer()
+                        this._complete_waiting_connect_ops(error)
+
+                    else:
+                        logger.info(
+                            "{}({}): State is {}.  Connection failed. Not triggering reconnection".format(
```

Comment:
Seems this log message isn't very useful re: state.

As I can tell, the only way this block triggers is if the state is in CONNECTED_OR_DISCONNECTED. So the state will ALWAYS be that in this case. Furthermore, connected and disconnected are pretty different. Is telling us that it is either connected or disconnected helpful?